prov/rxm/src/rxm_conn.c

```diff
@@ -1284,137 +1271,105 @@ static inline int rxm_conn_eq_event(struct rxm_ep *rxm_ep,
-		if (entry->rd < 0 &&
-		    entry->rd != -FI_ECONNREFUSED)
-			break;
-		if (rxm_conn_eq_event(rxm_ep, entry))
-			break;
-	}
-	return -1;
+	if (!entry->rd || entry->rd == -FI_EAGAIN)
+		return FI_SUCCESS;
+	if (entry->rd < 0 && entry->rd != -FI_ECONNREFUSED)
+		return entry->rd;
+
+	return rxm_conn_eq_event(rxm_ep, entry);
 }
 
-/* Atomic auto progress of EQ and CQ */
-static int rxm_conn_atomic_progress_eq_cq(struct rxm_ep *rxm_ep,
-					  struct rxm_msg_eq_entry *entry)
+static void *rxm_conn_atomic_progress(void *arg)
 {
-	struct rxm_fabric *rxm_fabric;
+	struct rxm_ep *ep = container_of(arg, struct rxm_ep, util_ep);
+	struct rxm_msg_eq_entry *entry;
+	struct rxm_fabric *fabric;
 	struct fid *fids[2] = {
-		&rxm_ep->msg_eq->fid,
-		&rxm_ep->msg_cq->fid,
+		&ep->msg_eq->fid,
+		&ep->msg_cq->fid,
 	};
 	struct pollfd fds[2] = {
 		{.events = POLLIN},
 		{.events = POLLIN},
 	};
-	int again;
 	int ret;
 
-	rxm_fabric = container_of(rxm_ep->util_ep.domain->fabric,
-				  struct rxm_fabric, util_fabric);
+	entry = alloca(RXM_MSG_EQ_ENTRY_SZ);
+	if (!entry)
+		return NULL;
+
+	fabric = container_of(ep->util_ep.domain->fabric,
+			      struct rxm_fabric, util_fabric);
 
-	ret = fi_control(&rxm_ep->msg_eq->fid, FI_GETWAIT, &fds[0].fd);
+	ret = fi_control(&ep->msg_eq->fid, FI_GETWAIT, &fds[0].fd);
 	if (ret) {
 		FI_WARN(&rxm_prov, FI_LOG_EP_CTRL,
-			"unable to get MSG EQ wait fd: %d\n", ret);
-		goto exit;
+			"unable to get msg EQ fd: %s\n", fi_strerror(ret));
+		return NULL;
 	}
 
-	ret = fi_control(&rxm_ep->msg_cq->fid, FI_GETWAIT, &fds[1].fd);
+	ret = fi_control(&ep->msg_cq->fid, FI_GETWAIT, &fds[1].fd);
 	if (ret) {
 		FI_WARN(&rxm_prov, FI_LOG_EP_CTRL,
-			"unable to get MSG CQ wait fd: %d\n", ret);
-		goto exit;
+			"unable to get msg CQ fd: %s\n", fi_strerror(ret));
+		return NULL;
 	}
 
-	memset(entry, 0, RXM_MSG_EQ_ENTRY_SZ);
-
-	while(1) {
-		ofi_ep_lock_acquire(&rxm_ep->util_ep);
-		again = fi_trywait(rxm_fabric->msg_fabric, fids, 2);
-		ofi_ep_lock_release(&rxm_ep->util_ep);
+	FI_INFO(&rxm_prov, FI_LOG_EP_CTRL, "Starting auto-progress thread\n");
+	while (ep->do_progress) {
+		/* TODO: Remove this lock, it can't protect anything */
+		ofi_ep_lock_acquire(&ep->util_ep);
+		ret = fi_trywait(fabric->msg_fabric, fids, 2);
+		ofi_ep_lock_release(&ep->util_ep);
 
-		if (!again) {
+		if (!ret) {
 			fds[0].revents = 0;
 			fds[1].revents = 0;
 
 			ret = poll(fds, 2, -1);
-			if (OFI_UNLIKELY(ret == -1)) {
-				if (errno == EINTR)
-					continue;
+			if (ret == -1 && errno != EINTR) {
 				FI_WARN(&rxm_prov, FI_LOG_EP_CTRL,
-					"Select error %d, closing CM thread\n",
-					errno);
-				goto exit;
+					"Select error %s, closing CM thread\n",
+					strerror(errno));
+				break;
 			}
 		}
-		if (again || fds[0].revents & POLLIN) {
-			if (rxm_conn_auto_progress_eq(rxm_ep, entry))
-				goto exit;
-		}
-		if (again || fds[1].revents & POLLIN)
-			rxm_ep->util_ep.progress(&rxm_ep->util_ep);
+		rxm_conn_auto_progress_eq(ep, entry);
```

Comment:
Question: Is the belief here that the latency of calling EQ progress is no worse than the previous check? Once connections are completed, EQ progress routine will be called when only the CQ needs processing.

Reply:
I'm more concerned that we miss an event and fail to progress the EQ or CQ.  One of the issues that's been reported is that apps are hanging, so this is more of a safety change.

The previous code had this behavior inside the while loop.  If trywait failed, we called progress.  Otherwise, we call a blocking poll call, then progress if needed.  This change removes the 'if needed' condition around progress.  If we unblock from poll, the code tries to progress anything that's ready.  We can prioritize progressing the EP over the EQ by flipping the 2 progress calls.